core/src/test/java/org/elasticsearch/repositories/blobstore/BlobStoreRepositoryRestoreTests.java

```diff
@@ -0,0 +1,277 @@
+                assertTrue("File [" + fileName + "] does not exist in store directory", directoryFiles.contains(fileName));
+                assertEquals(storeFile.length(), shard.store().directory().fileLength(fileName));
+            }
+        } finally {
+            if (shard != null && shard.state() != IndexShardState.CLOSED) {
+                try {
+                    shard.close("test", false);
+                } finally {
+                    IOUtils.close(shard.store());
+                }
+            }
+        }
+    }
+
+    /** Recover a shard from a snapshot using a given repository **/
+    private void recoverShardFromSnapshot(final IndexShard shard,
+                                          final Snapshot snapshot,
+                                          final BlobStoreRepository repository) throws IOException {
+        final Version version = Version.CURRENT;
+        final ShardId shardId = shard.shardId();
+        final String index = shard.shardId().getIndexName();
+        final IndexId indexId = new IndexId(index, UUIDs.randomBase64UUID());
+        final DiscoveryNode node = new DiscoveryNode(randomAlphaOfLength(25), buildNewFakeTransportAddress(), version);
+        final RecoverySource.SnapshotRecoverySource recoverySource = new RecoverySource.SnapshotRecoverySource(snapshot, version, index);
+        final ShardRouting shardRouting = newShardRouting(shardId, node.getId(), true, recoverySource, ShardRoutingState.INITIALIZING);
+
+        shard.markAsRecovering("from snapshot", new RecoveryState(shardRouting, node, null));
+        repository.restoreShard(shard, snapshot.getSnapshotId(), version, indexId, shard.shardId(), shard.recoveryState());
+    }
+
+    /** Snapshot a shard using a given repository **/
+    private void snapshotShard(final IndexShard shard,
+                                                   final Snapshot snapshot,
+                                                   final BlobStoreRepository repository) throws IOException {
+        final IndexShardSnapshotStatus snapshotStatus = new IndexShardSnapshotStatus();
+        try (Engine.IndexCommitRef indexCommitRef = shard.acquireIndexCommit(true)) {
+            Index index = shard.shardId().getIndex();
+            IndexId indexId = new IndexId(index.getName(), index.getUUID());
+
+            repository.snapshotShard(shard, snapshot.getSnapshotId(), indexId, indexCommitRef.getIndexCommit(), snapshotStatus);
+        }
+        assertEquals(IndexShardSnapshotStatus.Stage.DONE, snapshotStatus.stage());
+        assertEquals(shard.snapshotStoreMetadata().size(), snapshotStatus.numberOfFiles());
+        assertNull(snapshotStatus.failure());
+    }
+
+
+    /**
+     * A {@link BlobStoreRepository} implementation that works in memory.
+     *
+     * It implements only the methods required by the tests and is not thread safe.
+     */
+    class MemoryBlobStoreRepository extends BlobStoreRepository {
```

Comment:
I think we can just use an FsRepository for this. All our other shard-level tests do the same, so no need to optimize this. If we want to change that in the future, I think it's easier to switch to jimfs and continue using FsRepository.

Reply:
It's that, or we can replace replace FsRepository with this one, but we need to beef it up.

Reply:
Let's just use a FsRepository.